Adding MP version of Pourbaix Code in the pyEQL pourbaix branch

abhardwaj73 opened the pull request:
## Summary

Major changes:

- feature 1: Adding Pourbaix folder : `pourbaix_diagram` from `pymatgen.analysis` + compatibility and yaml files from `pymatgen.entries`
- feature 2: Adding test_pourbaix_diagram in the tests folder and adding import pymatgen, pyEQL. Also changing location of pourbaix_diagram file to local instead of `pymatgen.analysis`

## Todos

- fix 1: using `pip install -e .`, it gives error with pymatgen compatibility  `ERROR: No matching distribution found for pymatgen>=2024.9.10`
- fix 2 : after local install, need to update all locations in test file

@rkingsbury let me know if I'm on the right track

Also, I've cloned the main branch which was 13 commits ahead of the KingsburyLab/pourbaix branch. I thought it would help with pymatgen compatibility error, but it doesn't

rkingsbury commented:
Thanks @abhardwaj73 ! Yes, definitely on the right track. A few notes:

1. You did the right thing cloning `main`; I happened to make some updates that I had not synced back to this branch, but I've just merged a small PR to make `pourbaix` up to date with `main`, which makes your changes easier to see.
2. The `No matching distribution found` usually means that the version of python you're running does not support the version of a package you're trying to install. `pymatgen` currently supports 3.10+, so make sure you aren't on 3.9 or 3.8. For reference, pymatgen (and numpy and scipy and pyEQL and others) try to follow [numpy's deprecation policy](https://numpy.org/neps/nep-0029-deprecation_policy.html). According to this timetable, we're actually coming up on the time when python 3.10 support will be dropped, too.
3. You actually do not need the `yaml` files, because the only compatibilty class we need is `MPAqueousCompatibility` which does not read any information from those (so actually, you can remove the other compatibility classes from `compatibility.py` as well as the associated unit tests)
4. Please use a lowercase folder name `pourbaix` rather than `Pourbaix`, just for consistency

rkingsbury commented:
Also, remember to add `pymatgen.core.ion` and associated tests here as well

abhardwaj73 commented:
Thanks @rkingsbury , updating to latest python version helps in installing pyEQL locally. Added the mentioned files and removed the yaml files. 
Will update tests to make them pass, currently not passing.
- Doing pyEQL.pourbaix.pourbaix_diagram does not read my pourbaix_diagram file in the pyEQL, even though I have installed it using `pip install -e .`
- Are there other things I should keep in mind about changing the tests apart from the module address for importing

rkingsbury commented on pourbaix/pourbaix_diagram.py:
```diff
@@ -0,0 +1,1138 @@
+from pymatgen.analysis.reaction_calculator import Reaction, ReactionError
+from pymatgen.core import Composition, Element
+from pymatgen.core.ion import Ion
+from pymatgen.entries.compatibility import MU_H2O
```
This can change to `from pyEQL.pourbaix.compatibility import...`

rkingsbury commented on pourbaix/pourbaix_diagram.py:
```diff
@@ -0,0 +1,1138 @@
+from pymatgen.analysis.phase_diagram import PDEntry, PhaseDiagram
+from pymatgen.analysis.reaction_calculator import Reaction, ReactionError
+from pymatgen.core import Composition, Element
+from pymatgen.core.ion import Ion
```
here you can also move the import `from pyEQL.pourbaix.ion import Ion`

rkingsbury commented:
> Doing pyEQL.pourbaix.pourbaix_diagram does not read my pourbaix_diagram file in the pyEQL

You mean `from pyEQL.pourbaix.pourbaix_diagram import XXXX` fails?

> Are there other things I should keep in mind about changing the tests apart from the module address for importing

You'll want to remove tests of the other compatibility classes (MIT, MP, MP2020) because they load the yaml files. You can also remove those classes, and the associated imports, from `compatibility.py`.

rkingsbury commented on tests/test_pourbaix_diagram.py:
```diff
@@ -0,0 +1,356 @@
+from monty.serialization import dumpfn, loadfn
+from pytest import approx
+
+from pyEQL.Pourbaix.pourbaix_diagram import (
+    IonEntry,
+    MultiEntry,
+    PourbaixDiagram,
+    PourbaixEntry,
+    PourbaixPlotter,
+    ion_or_solid_comp_object,
+)
```
looks like you've got a casing error in the import statement. `Pourbaix` should be `pourbaix`

rkingsbury commented on tests/test_pourbaix_diagram.py:
```diff
@@ -0,0 +1,356 @@
+    ion_or_solid_comp_object,
+)
+from pymatgen.core.composition import Composition
+from pymatgen.core.ion import Ion
```
switch this import to pyEQL

rkingsbury commented on tests/test_pourbaix_diagram.py:
```diff
@@ -0,0 +1,356 @@
+from pymatgen.entries.computed_entries import ComputedEntry
+from pymatgen.util.testing import TEST_FILES_DIR, PymatgenTest
+
+TEST_DIR = f"{TEST_FILES_DIR}/analysis/pourbaix_diagram"
+
```
It looks like this test calls some additional files from the pymatgen TEST_FILES_DIR, so those files will need to be copied into our tests directory, and this path updated

rkingsbury commented:
Note that when you run `pytest`, you can specifically call it on only particular tests by doing

`pytest test_mod.py::TestClass::test_method`

This is useful in cases like yours, because right now there a bunch of tests that you probably expect to fail (e.g., ones that require yaml files), so it can be helpful to just test a few at a time to get the basics working, before addressing others.

See more at https://docs.pytest.org/en/7.1.x/how-to/usage.html#specifying-which-tests-to-run

abhardwaj73 commented:
> You'll want to remove tests of the other compatibility classes (MIT, MP, MP2020) because they load the yaml files. You can also remove those classes, and the associated imports, from `compatibility.py`.

Removing the following classes:

`MaterialsProject2020Compatibility`
`MaterialsProjectCompatibility`
`MITAqueousCompatibility`
`MITCompatibility`

Only keeping : `MaterialsProjectAqueousCompatibility`

abhardwaj73 commented:
Changes in this version:
- Moved pourbaix inside src/pyEQL
- added pourbaix branch inside .github/workflows/testing.yaml (same as main branch update by @rkingsbury)
- Added back `MP2020Compatibility` and its file, since it was needed for `MaterialsProjectAqueousCompatibility`
- updated `TEST_DIR = f"{TEST_FILES_DIR}/analysis/pourbaix_diagram"` to the current test directory and added the `pourbaix_test_data.json`

Testing:
- `test_pourbaix_diagram` `test_ion` `test_compatibility` were added and all of them passed
- Testing all the tests: 1 failed, 174 passed, 3 xfailed
1 failed : FAILED test_solution.py::test_from_preset - FileNotFoundError: [Errno 2] No such file or directory: 'src/pyEQL/presets/seawater.yaml'
3 XFAIL: XFAIL test_activity.py::test_water_activity_pitzer_nacl_1
XFAIL test_dielectric.py::Test_dielectric::test_dielectric_constant8
XFAIL test_effective_pitzer.py::Test_effective_pitzer::test_effective_pitzer_na2so4_activity

rkingsbury commented:
Thanks @abhardwaj73 ! Tagging @SuixiongTay as well as an FYI.
